src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs

```diff
@@ -1677,36 +1676,80 @@ public void VerifyXmlResolver(bool provideResolver)
+                Action requestReceived,
                 CancellationToken cancellationToken)
             {
+                const string Response =
+                    "HTTP/1.1 200 OK\r\nContent-Type: text/plain; charset=UTF-8\r\nContent-Length: 0\r\n\r\n";
+
                 while (!cancellationToken.IsCancellationRequested)
                 {
-                    HttpListenerContext ctx;
+                    Socket socket;
 
                     try
                     {
-                        ctx = await listener.GetContextAsync();
+#if NETCOREAPP
+                        socket = await listener.AcceptSocketAsync(cancellationToken);
+#else
+                        socket = await listener.AcceptSocketAsync();
+#endif
                     }
                     catch
                     {
                         break;
                     }
 
-                    HttpListenerRequest req = ctx.Request;
-                    requestReceived(req);
-
-                    using (HttpListenerResponse resp = ctx.Response)
+                    using (socket)
                     {
-                        resp.ContentType = "text/plain";
-                        resp.ContentEncoding = Encoding.UTF8;
-                        resp.ContentLength64 = 0;
+                        requestReceived();
+                        byte[] buf = new byte[1024];
+
+#if NETCOREAPP
+                        // The CoreFX XML libraries throw an exception if we don't drain the request
+                        // (unless we don't dispose the socket).
+                        // For some reason each ReceiveAsync only gets 10 bytes at a time,
+                        // so just keep reading.
+                        int offset = 0;
+
+                        do
+                        {
+                            int read = await socket.ReceiveAsync(buf.AsMemory(offset), cancellationToken);
+
+                            if (read <= 0)
+                            {
+                                break;
+                            }
+
+                            offset += read;
+
+                            if (offset >= buf.Length)
+                            {
+                                throw new InvalidOperationException();
+                            }
+
+                            if (offset > sizeof(int))
+                            {
+                                int last = BinaryPrimitives.ReadInt32LittleEndian(buf.AsSpan(offset - sizeof(int)));
+
+                                // CRLFCRLF, little-endian
+                                if (last == 0x0A0D0A0D)
+                                {
+                                    break;
+                                }
+                            }
+                        } while (true);
+
+                        int len = Encoding.ASCII.GetBytes(Response, buf);
+                        await socket.SendAsync(buf.AsMemory(0, len), cancellationToken);
+#else
```

Comment:
Probably something to do with HttpWebRequest vs HttpClient on the part of the caller.

It boils down to:
* If I don't dispose the Socket, I don't need to read on corefx.
* If I do dispose the Socket, but don't do this reading, the caller side in corefx throws that the connection was closed before it finishes writing.
* I also had a read in the netfx codepath, but it wasn't necessary to make the test pass, so I removed it.  (It was complicated, because Socket pre-net6 is complicated)

Reply:
Do you expect more requests on the same connection? If not, I think you should call `Shutdown(SocketShutdown.Both)`. Without it data are not guarantied to be flushed and written on disposal. 
Alternatively you can use NetworkStream.  More useable abstraction and it already does it

https://github.com/dotnet/runtime/blob/32ac664fca93cb692597e75021ae5e8ee7b75b1c/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs#L380-L395